File: backend/GameApp.Domain/ValueObjects/Items/ItemName.cs
namespace GameApp.Domain.ValueObjects.Items;


// Value Object for a game's inventory item name
public class ItemName
{
    private static readonly string _messageIfEmpty = "Item name cannot be empty";
    
    private readonly string Name;

    public ItemName(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ArgumentException(_messageIfEmpty, nameof(value));

        Name = value;
    }

    public string GetName() => Name;

    public ItemName SetName(string newName)
    {
        if (string.IsNullOrWhiteSpace(newName))
            throw new ArgumentException(_messageIfEmpty, nameof(newName));

        return new ItemName(newName);
    }

    public override bool Equals(object? obj)
    {
        if (obj is not ItemName other)
            return false;

        return Name == other.Name;
    }

    public override int GetHashCode() => Name.GetHashCode();

    public override string ToString() => Name;
}

File: frontend/gameapp-frontend/src/domain/value-objects/items/item-name.ts
// Value Object for item name
export class ItemName {
    static ERROR_EMPTY = "Item name cannot be empty";

    private _name: string;

    constructor(name: string) {
        this._name = name;
    }

    get name(): string {
        return this._name;
    }

    setName(newName: string): ItemName {
        return new ItemName(newName);
    }

    equals(other: unknown): boolean {
        return other instanceof ItemName &&
               this._name === other._name;
    }

    toString(): string {
        return this._name;
    }
}
